reward/reward_distributor.go

```diff
@@ -38,113 +50,469 @@ type governanceHelper interface {
-		// magma hardfork
-		return totalGasUsed.Mul(totalGasUsed, header.BaseFee)
+func NewRewardConfig(header *types.Header, config *params.ChainConfig) (*rewardConfig, error) {
+	if config.Governance == nil {
+		return nil, errors.New("no config.Governance")
+	}
+	if config.Governance.Reward == nil {
+		return nil, errors.New("no config.Governance.Reward")
+	}
+
+	rules := config.Rules(header.Number)
+
+	cnRatio, kgfRatio, kirRatio, totalRatio, err := parseRewardRatio(config.Governance.Reward.Ratio)
+	if err != nil {
+		return nil, err
+	}
+
+	var cnProposerRatio, cnStakingRatio, cnTotalRatio int64
+	if rules.IsKore {
+		cnProposerRatio, cnStakingRatio, cnTotalRatio, err = parseRewardKip82Ratio(config.Governance.Reward.Kip82Ratio)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &rewardConfig{
+		// hardfork rules
+		rules: rules,
+
+		// values calculated from block header
+		totalFee: GetTotalTxFee(header, config),
+
+		// values from ChainConfig
+		mintingAmount: new(big.Int).Set(config.Governance.Reward.MintingAmount),
+		minimumStake:  new(big.Int).Set(config.Governance.Reward.MinimumStake),
+		deferredTxFee: config.Governance.Reward.DeferredTxFee,
+
+		// parsed ratio
+		cnRatio:    big.NewInt(cnRatio),
+		kgfRatio:   big.NewInt(kgfRatio),
+		kirRatio:   big.NewInt(kirRatio),
+		totalRatio: big.NewInt(totalRatio),
+
+		// parsed KIP82 ratio
+		cnProposerRatio: big.NewInt(cnProposerRatio),
+		cnStakingRatio:  big.NewInt(cnStakingRatio),
+		cnTotalRatio:    big.NewInt(cnTotalRatio),
+	}, nil
+}
+
+func GetTotalTxFee(header *types.Header, config *params.ChainConfig) *big.Int {
+	totalFee := new(big.Int).SetUint64(header.GasUsed)
+	if config.IsMagmaForkEnabled(header.Number) {
+		totalFee = totalFee.Mul(totalFee, header.BaseFee)
 	} else {
-		return totalGasUsed.Mul(totalGasUsed, rewardConfig.unitPrice)
+		totalFee = totalFee.Mul(totalFee, new(big.Int).SetUint64(config.UnitPrice))
 	}
+	return totalFee
 }
 
-func (rd *RewardDistributor) txFeeBurning(txFee *big.Int) *big.Int {
-	return txFee.Div(txFee, big.NewInt(2))
+// GetBlockReward returns the actual reward amounts paid in this block
+// Used in klay_getReward RPC API
+func GetBlockReward(header *types.Header, config *params.ChainConfig) (*RewardSpec, error) {
+	var spec *RewardSpec
+	var err error
+
+	if config.Istanbul == nil {
+		return nil, errors.New("no IstanbulConfig")
+	}
+
+	policy := config.Istanbul.ProposerPolicy
+	if policy == uint64(istanbul.RoundRobin) || policy == uint64(istanbul.Sticky) {
+		spec, err = CalcDeferredRewardSimple(header, config)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		spec, err = CalcDeferredReward(header, config)
+		if err != nil {
+			return nil, err
+		}
+
+		// Compensate the difference between CalcDeferredReward() and actual payment.
+		// If not DeferredTxFee, CalcDeferredReward() assumes 0 total_fee, but
+		// some non-zero fee is paid to the proposer.
+		if !config.Governance.Reward.DeferredTxFee {
+			var blockFee *big.Int
```

Comment:
The difference comes from https://github.com/klaytn/klaytn/blob/v1.9.1/blockchain/state_transition.go#L288, which does not care about magma burn if `not DeferredTxFee`. Therefore the compensation here must not care about magma burn.

```suggestion
			blockFee := getTotalTxFee(header, config)
```

Reply:
fixed in cddc37b2